### ISSUE TITLE
Installation on RedHat (RHEL 7)

### ISSUE THREAD

**Issue by snowman2** [tethysplatform/tethys#137]
# Installation on RedHat (RHEL 7)
Note: Docker instructions are NOT included. They may be added in the future.

If you have problems or random errors, see Troubleshooting section at the bottom!
Also, check error logs if you are far enough: 
```
$ tail /var/log/httpd/tethys_default.error.log
```
## Step 1: Install the Dependencies

```
$ sudo yum update 
$ sudo yum install python-devel python-virtualenv postgresql-devel libxml2-devel libxslt-devel libffi-devel git-core gcc openssl-devel
# easy_install pip
```
## Step 2: Install Postgresql
Due to the headache caused by installing the docker in RHEL 7, it was way easier to install postgresql than use the docker.

### If you want to attempt to use docker:
This is as far as I got. Good luck!
```
$ vim /etc/yum.repos.d/docker.repo
```
Add this text:
[dockerrepo]
name=Docker Repository
baseurl=https://yum.dockerproject.org/repo/main/centos/7
enabled=1 
gpgcheck=1
gpgkey=https://yum.dockerproject.org/gpg
```
# yum install docker-engine
# sudo gpasswd -a ${USER} docker
# newgrp
# systemctl enable docker
```
Start docker:
```
# firewall-cmd --zone=public --permanent --add-port=5435/tcp 
# firewall-cmd --reload
# systemctl start docker
```
See steps 2-5: 
http://docs.tethysplatform.org/en/latest/installation/linux.html

### Install Postgresql Without Docker  (skipping steps 2,4,5 in docs)

PostgreSQL  `PostgreSQL database system, version 9.0 or higher (9.3 recommended). <http://www.postgresql.org/download/>`_
PostGIS 2.1   `PostGIS spatial extension for PostgreSQL, version 2.1 or higher. <http://postgis.net/install>`_

See: https://wiki.postgresql.org/wiki/YUM_Installation
```
$ yum install postgresql-server postgis postgis-utils postgis-docs
$ postgresql-setup initdb
$ systemctl enable postgresql
$ systemctl start postgresql
```
#### Create Database and Database Users
-------------------------------------

Create three database users with databases. You will be prompted to create passwords for each user. Take note of the passwords, because you will need to use them in the next step. To do so, run the following commands in the terminal:

```
$ sudo su - postgres 
$ createuser -S -D -R -P tethys_default 
$ createdb -O tethys_default tethys_default -E utf-8 
$ createuser -S -d -R -P tethys_db_manager 
$ createdb -O tethys_db_manager tethys_db_manager -E utf-8 
$ createuser --superuser -d -R -P tethys_super 
$ createdb -O tethys_super tethys_super -E utf-8 
$ exit
```

Tip:

    If you would like to have access to the Tethys databases through a graphical user interface such as
    `pgAdmin III <http://www.pgadmin.org/>`_, use the **tethys_super** database user and password as credentials
    for the connection.

## Step 3: Create Virtual Environment and Install Tethys Platform
See: http://docs.tethysplatform.org/en/latest/installation/linux.html#create-virtual-environment-and-install-tethys-platform

## Step 6: Create Settings File and Configure Settings

a.	PORT for postgresql is 5432 by default (http://www.postgresql.org/docs/9.1/static/runtime-config-connection.html)

## Steps 6(b-e)-10:
See: http://docs.tethysplatform.org/en/latest/installation/linux.html

# Production Installation for RedHat
http://docs.tethysplatform.org/en/latest/production/installation.html

## 1. Install Tethys Portal
See above for instructions for RedHat
## 2. Install Apache and Dependencies
```
$  sudo yum install httpd mod_wsgi vim
$ systemctl start httpd
$ systemctl enable httpd
```
## Steps 3-12
See: http://docs.tethysplatform.org/en/latest/production/installation.html

## 13. Transfer Ownership to Apache
```
$ sudo chown -R apache:apache /usr/lib/tethys/src /var/www/tethys/static /var/www/.tethyscluster
```
## 14. Enable Site and Restart Apache
Listen for other ports:
```
$ vim /etc/httpd/conf/httpd.conf
```
#Change this to Listen on specific IP addresses as shown below to 
#prevent Apache from glomming onto all bound IP addresses.
#
Listen 12.34.56.78:80

Open Ports in Firewall:
https://docs.fedoraproject.org/en-US/Fedora/19/html/Security_Guide/sec-Open_Ports_in_the_firewall-CLI.html
```
$ firewall-cmd --zone=public --permanent --add-port=80/tcp
$ systemctl restart httpd
```
# Troubleshooting:

## SELinux Troubleshooting
See: https://github.com/tethysplatform/tethys/issues/136

## django.db.utils.OperationalError: FATAL:  Ident authentication failed for user "tethys_default"

http://www.cyberciti.biz/faq/psql-fatal-ident-authentication-failed-for-user/
```
$ vim /var/lib/pgsql/data/pg_hba.conf
```
```
local   all         postgres                          ident
local   all         all                               allow 
#IPv4 local connections:
host    all         all         127.0.0.1/32          allow 
#IPv6 local connections:
host    all         all         ::1/128               allow
```
```
$ systemctl restart postgresql
```
## https://groups.google.com/forum/#!topic/tethysplatform/s8Mba8P4qkg

If you see:

$ tail /var/log/httpd/tethys_default.error.log 
os.mkdir(TETHYSCLUSTER_CFG_DIR)
OSError: [Errno 13] Permission denied: '/usr/share/httpd/.tethyscluster'

Try:
```
$ mkdir /usr/share/httpd/.tethyscluster/
$ touch /usr/share/httpd/.tethyscluster/aws_config
$ touch /usr/share/httpd/.tethyscluster/azure_config
$ touch /usr/share/httpd/.tethyscluster/config
$ touch /usr/share/httpd/.tethyscluster/tethys_config
```
Here is the result after permissions added:
```
$ ls -lah /usr/share/httpd/.tethyscluster/
drwsrwsr-x. 2 apache apache 75 Sep 22 12:25 .
-rw-rw-r--. 1 apache apache  0 Sep 22 12:23 aws_config
-rw-rw-r--. 1 apache apache  0 Sep 22 12:25 azure_config
-rw-rw-r--. 1 apache apache  0 Sep 22 12:18 config
-rw-rw-r--. 1 apache apache  0 Sep 22 12:22 tethys_config
```

**Comment by snowman2** [comment]
# Install CKAN on RedHat (RHEL 7)
Instructions based on:
- https://github.com/ckan/ckan/wiki/How-to-install-CKAN-2.0-on-CentOS-7
- http://docs.ckan.org/en/latest/maintaining/installing/install-from-source.html
- https://github.com/okfn/ckan-barnet/wiki/How-To-Install-Apache-Solr-on-Red-Hat

## 1. Install the required packages
Install and activate the Extra Packages for Enterprise Linux (EPEL) Repository (it may already be installed)
```
# rpm -Uvh http://dl.fedoraproject.org/pub/epel/7/x86_64/e/epel-release-7.5.noarch.rpm
```
Install packages:
```
# yum install xml-commons git subversion mercurial postgresql-server postgresql-devel \
postgresql python-devel libxslt libxslt-devel libxml2 libxml2-devel python-virtualenv \
gcc gcc-c++ make java-1.8.0-openjdk-devel java-1.8.0-openjdk tomcat tomcat-webapps \
tomcat-admin-webapps xalan-j2 unzip policycoreutils-python mod_wsgi httpd
```
## 2. Install CKAN

First, create a CKAN User. The ckan user is created with a shell of /sbin/nologin and a home directory of /usr/lib/ckan to mirror what is shown in the CKAN Deployment documentation.
```
# useradd -m -s /sbin/nologin -d /usr/lib/ckan -c "CKAN User" ckan
```
Open the newly created directory up for read access so that the content will eventually be able to be served out via httpd.
```
# chmod 755 /usr/lib/ckan
```
Switch to the ckan user.
```
# su -s /bin/bash - ckan
```
Install an isolated Python environment, called default, to host CKAN from.
```
# virtualenv --no-site-packages default
```
Activate the newly installed Python environment.
```
# . default/bin/activate
```
Download and install version 2.4.1 of CKAN.
```
# pip install -e 'git+https://github.com/ckan/ckan.git@ckan-2.4.1#egg=ckan'
```
Download and install the necessary Python modules to run CKAN into the isolated Python environment
```
# pip install -r /usr/lib/ckan/default/src/ckan/requirements.txt
```
Return back to root user by typing exit or pressing Ctrl + D.

## 3. Configure PostgreSQL
```
# postgresql-setup initdb
# systemctl enable postgresql
```
Edit /var/lib/pgsql/data/pg_hba.conf so it will accept passwords for login while still allowing the local postgres user to manage via ident login. The relevant changes to pg_hba.conf are as follows:
```
local   all         postgres                          ident
local   all         all                               allow 
#IPv4 local connections:
host    all         all         127.0.0.1/32          allow 
#IPv6 local connections:
host    all         all         ::1/128               allow
```
Start PostgreSQL
```
# systemctl start postgresql.service
```
Switch to postgres user
```
# su - postgres
```
List existing databases:
```
$ psql -l
```
Check that the encoding of databases is UTF8, if not internationalisation may be a problem. Since changing the encoding of PostgreSQL may mean deleting existing databases, it is suggested that this is fixed before continuing with the CKAN install.

Next you’ll need to create a database user if one doesn’t already exist. Create a new PostgreSQL database user called ckan_default, and enter a password for the user when prompted. You’ll need this password later
```
$ createuser -S -D -R -P ckan_default
```
Create a new PostgreSQL database, called ckan_default, owned by the database user you just created.
```
$ createdb -O ckan_default ckan_default -E utf-8
```
Exit the postgres user environment with Ctrl + D or exit

## 4. Create a CKAN Configuration

Switch back to root user and create a directory to contain the site’s config files:
```
# mkdir -p /etc/ckan/default
# chown -R ckan /etc/ckan/
```
Switch to ckan user and create a CKAN config file:
```
$ su -s /bin/bash - ckan
$ . default/bin/activate
$ cd /usr/lib/ckan/default/src/ckan
$ paster make-config ckan /etc/ckan/default/development.ini
```
Edit the development.ini file in a text editor, changing the following options:
```
sqlalchemy.url = postgresql://ckan_default:pass@localhost/ckan_default
ckan.site_id = default
solr_url = http://127.0.0.1:8080/solr/ckan_default
```
Exit from running as the ckan user with Ctrl + D or exit.

## 5. Setup Apache SOLR
Download and extract Apache SOLR
```
# curl http://archive.apache.org/dist/lucene/solr/4.2.1/solr-4.2.1.tgz | tar xzf -
```
Copy the Apache SOLR war to the desired location.
```
# cp solr-4.2.1/dist/solr-4.2.1.war /usr/share/solr/solr.war
```

Copy the example Apache SOLR configuration to the ckan_default directory.
This is the directory where Solr will store the search index files for our core.
Create the directory /etc/solr/ckan_default, and copy the apache-solr-4.0.0/example/solr/conf directory into it:
mkdir -p /etc/solr/ckan_default
cp -r solr-4.2.1/example/solr/collection1/conf /etc/solr/ckan_default

1. Create the file /usr/share/solr/solr.xml, with the following contents:
```
<solr persistent="true" sharedLib="lib">
    <cores adminPath="/admin/cores">
        <core name="ckan_default" instanceDir="ckan_default">
            <property name="dataDir" value="/var/lib/solr/data/ckan_default" />
        </core>
    </cores>
</solr>
```
This file lists the different Solr cores, in this example we have just a single core called ckan_default.
Create the data directory for your Solr core, run this command in a terminal:
```
# mkdir –p /var/lib/solr/data/ckan_default
```
This directory holds the configuration files for your Solr core.
Replace the /etc/solr/ckan_default/schema.xml file with a symlink to CKAN’s schema.xml file:
```
# mv /etc/solr/ckan_default/conf/schema.xml /etc/solr/ckan_default/conf/schema.xml.bak
# ln -s /usr/lib/ckan/default/src/ckan/ckan/config/solr/schema.xml /etc/solr/ckan_default/conf/schema.xml
```
Edit /etc/solr/ckan_default/conf/solrconfig.xml and change the <dataDir> tag to this:
```
<dataDir>${dataDir}</dataDir>
```
This configures our ckan_default core to use the data directory you specified for it in solr.xml.
Create the directory /usr/share/solr/ckan_default and put a symlink to the conf directory in it:
```
# mkdir /usr/share/solr/ckan_default
# ln -s /etc/solr/ckan_default/conf /usr/share/solr/ckan_default/conf
```
Create a new file, called /etc/tomcat/Catalina/localhost/solr.xml, with the following contents:
```
<Context docBase="/usr/share/solr/solr.war" debug="0" privileged="true" allowLinking="true" crossContext="true">
    <Environment name="solr/home" type="java.lang.String" value="/usr/share/solr" override="true" />
</Context>
```
Change the solr_url setting in your /etc/ckan/default/development.ini or /etc/ckan/default/production.ini file to point to your new Solr core, for example:
```
solr_url = http://127.0.0.1:8080/solr/ckan_default
```
Set Permissions
Make tomcat the owner of the Solr directories.
```
# chown -R tomcat:tomcat /usr/share/solr /var/lib/solr 
```
Enable Tomcat
Configure Tomcat to start on system boot.
```
# systemctl enable tomcat.service 
```
Start Tomcat
```
# systemctl start tomcat.service 
```
If Tomcat installation was successfull, you may find its web interface at:
 http://www.yourdomain.com:8080/ 
If Apache Solr installation was successfull, you may find its web interface at:
 http://www.yourdomain.com:8080/solr

## 6. Create the Database Tables
Switch back to running as the ckan user, activate the isolated Python environment, and change to the CKAN source directory.
```
# su -s /bin/bash - ckan . default/bin/activate cd default/src/ckan 
```
Initialize the CKAN database.
```
$ paster db init -c /etc/ckan/default/development.ini 
```
You may see a few errors but then Initialising DB: SUCCESS.

##  Setup the Datastore (Optional)
Follow the instructions in Setting up the DataStore to create the required databases and users, set the right permissions and set the appropriate values in your CKAN config file.
Note: Setting up the DataStore is optional.
## Link to who.ini
You should still be in the python virtualenv for this step, if not, do the following:
```
# su -s /bin/bash - ckan . default/bin/activate cd default/src/ckan 
```
who.ini (the Repoze.who configuration file) needs to be accessible in the same directory as your CKAN config file, so create a symlink to it:
```
$ ln -s /usr/lib/ckan/default/src/ckan/who.ini /etc/ckan/default/who.ini 
```
## Create a WSGI file
Create your site’s WSGI script file /etc/ckan/default/apache.wsgi with the following contents:
```
import os
activate_this = os.path.join('/usr/lib/ckan/default/bin/activate_this.py')
execfile(activate_this, dict(__file__=activate_this))

from paste.deploy import loadapp
config_filepath = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'development.ini')
from paste.script.util.logging_config import fileConfig
fileConfig(config_filepath)
application = loadapp('config:%s' % config_filepath)
```
The modwsgi Apache module will redirect requests to your web server to this WSGI script file. The script file then handles those requests by directing them on to your CKAN instance (after first configuring the Python environment for CKAN to run in).
Exit the ckan user with Ctrl + D or exit.
10. Create the Apache config file
Create your site’s Apache config file at /etc/httpd/conf.d/ckan_default.conf, with the following contents:
```
WSGISocketPrefix /var/run/wsgi
<VirtualHost 0.0.0.0:80>
    ServerName default.yourdomain.com
    ServerAlias www.default.yourdomain.com
    WSGIScriptAlias / /etc/ckan/default/apache.wsgi

    # Pass authorization info on (needed for rest api).
    WSGIPassAuthorization On

    # Deploy as a daemon (avoids conflicts between CKAN instances).
    WSGIDaemonProcess ckan_default display-name=ckan_default processes=2 threads=15

    WSGIProcessGroup ckan_default

    # Add this to avoid Apache show error: 
    # "AH01630: client denied by server configuration: /etc/ckan/default/apache.wsgi" 
    <Directory /etc/ckan/default>
    Options All
    AllowOverride All
    Require all granted
    </Directory>

    ErrorLog /var/log/httpd/ckan_default.error.log
    CustomLog /var/log/httpd/ckan_default.custom.log combined
</VirtualHost>
```
Replace default.ckanhosted.com and www.default.ckanhosted.com with the domain name for your site.
This tells the Apache modwsgi module to redirect any requests to the web server to the WSGI script that you created above. Your WSGI script in turn directs the requests to your CKAN instance.

## Configure Apache
Enable httpd to start on system boot
```
# systemctl enable httpd 
```
Listen for other ports:
```
$ vim /etc/httpd/conf/httpd.conf
```
#Change this to Listen on specific IP addresses as shown below to 
#prevent Apache from glomming onto all bound IP addresses.
#
Listen 12.34.56.78:80

Open Ports in Firewall:
https://docs.fedoraproject.org/en-US/Fedora/19/html/Security_Guide/sec-Open_Ports_in_the_firewall-CLI.html
```
$ firewall-cmd --zone=public --permanent --add-port=80/tcp
$ systemctl start httpd
```

**Comment by snowman2** [comment]
# Install GeoServer 2.7.2 on RedHat (RHEL 7):
Note: Without docker because it is lighter weight and less of a headache to get it to work.
Docs based on:
- http://docs.geoserver.org/2.7.2/user/installation/war.html

## Install Dependencies:
```
# yum install tomcat vim
```
Download and install oracle java 7:
```
# wget --no-cookies --no-check-certificate --header "Cookie: gpw_e24=http%3A%2F%2Fwww.oracle.com%2F; oraclelicense=accept-securebackup-cookie" "http://download.oracle.com/otn-pub/java/jdk/7u79-b15/jdk-7u79-linux-x64.rpm"
# yum localinstall jdk-7u79-linux-x64.rpm
# rm jdk-7u79-linux-x64.rpm
```
## Install geoserver:
```
# mkdir /usr/local/geoserver
# cd /usr/local/geoserver
# wget http://sourceforge.net/projects/geoserver/files/GeoServer/2.7.2/geoserver-2.7.2-war.zip
# unzip geoserver-2.7.2-war.zip
# rm geoserver-2.7.2-war.zip
# cp /usr/local/geoserver/geoserver.war /usr/share/tomcat/webapps
# systemctl enable tomcat
# systemctl start tomcat
```
## Enable JSONP:
```
vim /usr/share/tomcat/webapps/geoserver/WEB-INF/web.xml
```
Uncomment JSONP:
``` 
 <!--When true the JSONP (text/javascript) output format is enabled -->
  <!--
  <context-param>
    <param-name>ENABLE_JSONP</param-name>
    <param-value>true</param-value>
  </context-param>
  -->
```
Should look like:
```
  <!--When true the JSONP (text/javascript) output format is enabled -->
  <context-param>
    <param-name>ENABLE_JSONP</param-name>
    <param-value>true</param-value>
  </context-param>
```
Restart tomcat server:
```
# systemctl restart tomcat
```

**Comment by sdc50** [comment]
This issue will be addressed by issue #146
